### ISSUE TITLE
ERROR: null value in column "database_type" violates not-null constraint

### ISSUE THREAD

**Issue by stieg** [metabase/metabase#9888]
Hit this bug while setting up Metabase.  

**Requested details:**
* Database type being synchronized: MongoDB 4.0.6
* Metabase Version: v0.32.5
* Metabase Hosting Environment: Docker on ECS
* Metabase internal DB: Postgres 9.6

**Stack:**
```
May 04 23:50:00 DEBUG metabase.sync.util :: STARTING: step 'sync-fields' for mongo Database 4 'Foo DB'
May 04 23:51:09 DEBUG metabase.sync.sync-metadata.fields :: Hash of Table 45 'images' matches stored hash, skipping Fields sync
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:23 WARN metabase.driver.common :: Don't know how to map class 'class org.bson.BsonUndefined' to a Field base_type, falling back to :type/*.
May 04 23:51:24 INFO metabase.sync.sync-metadata.fields.sync-metadata :: Database type of Table 46 'bookings' Field 'on' has changed from 'NULL' to 'null'.
May 04 23:51:24 ERROR metabase.sync.util :: Error syncing Fields for Table 'Table 46 'bookings'': Batch entry 0 UPDATE "metabase_field" SET "database_type" = NULL, "updated_at" = '2019-05-04 23:51:24.019-05' WHERE "id" = 511 was aborted: ERROR: null value in column "database_type" violates not-null constraint
Detail: Failing row contains (511, 2019-05-02 23:45:33.076-05, 2019-05-04 23:51:24.019-05, on, type/*, null, t, null, t, 0, 46, 461, on, normal, null, 2019-05-02 23:46:22.103-05, null, null, {"global":{"distinct-count":1,"nil%":1.0}}, 4, null, auto-list, null). Call getNextException to see other errors in the batch.
("org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:148)"
"org.postgresql.core.ResultHandlerDelegate.handleError(ResultHandlerDelegate.java:50)"
"org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2184)"
"org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:481)"
"org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:840)"
"org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1538)"
"com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeBatch(NewProxyPreparedStatement.java:2437)"
"clojure.java.jdbc$execute_batch.invokeStatic(jdbc.clj:591)"
"clojure.java.jdbc$execute_batch.invoke(jdbc.clj:584)"
"clojure.java.jdbc$db_do_execute_prepared_statement$fn__14304.invoke(jdbc.clj:1039)"
"clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:799)"
"clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)"
"clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:782)"
"clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)"
"clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:1038)"
"clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:1024)"
"clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:1062)"
"clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:1042)"
"clojure.java.jdbc$execute_BANG_$execute_helper__14372.invoke(jdbc.clj:1446)"
"clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1450)"
"clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1417)"
"toucan.db$execute_BANG_.invokeStatic(db.clj:411)"
"toucan.db$execute_BANG_.doInvoke(db.clj:406)"
"clojure.lang.RestFn.invoke(RestFn.java:410)"
"toucan.db$update_BANG_.invokeStatic(db.clj:466)"
"toucan.db$update_BANG_.invoke(db.clj:455)"
"toucan.db$update_BANG_.invokeStatic(db.clj:475)"
"toucan.db$update_BANG_.invoke(db.clj:455)"
"--> sync.sync_metadata.fields.sync_metadata$fn__46957$update_field_metadata_if_needed_BANG___46962$fn__46963.invoke(sync_metadata.clj:74)"
"sync.sync_metadata.fields.sync_metadata$fn__46957$update_field_metadata_if_needed_BANG___46962.invoke(sync_metadata.clj:19)"
"sync.sync_metadata.fields.sync_metadata$fn__47016$update_metadata_BANG___47021$fn__47022$iter__47023__47027$fn__47028.invoke(sync_metadata.clj:99)"
"sync.util$sum_for_STAR_.invokeStatic(util.clj:462)"
"sync.util$sum_for_STAR_.invoke(util.clj:459)"
"sync.sync_metadata.fields.sync_metadata$fn__47016$update_metadata_BANG___47021$fn__47022.invoke(sync_metadata.clj:95)"
"sync.sync_metadata.fields.sync_metadata$fn__47016$update_metadata_BANG___47021.invoke(sync_metadata.clj:89)"
"sync.sync_metadata.fields.sync_metadata$fn__46991$update_nested_fields_metadata_BANG___46996$fn__46997.invoke(sync_metadata.clj:86)"
"sync.sync_metadata.fields.sync_metadata$fn__46991$update_nested_fields_metadata_BANG___46996.invoke(sync_metadata.clj:80)"
"sync.sync_metadata.fields.sync_metadata$fn__47016$update_metadata_BANG___47021$fn__47022$iter__47023__47027$fn__47028.invoke(sync_metadata.clj:100)"
"sync.util$sum_for_STAR_.invokeStatic(util.clj:462)"
"sync.util$sum_for_STAR_.invoke(util.clj:459)"
"sync.sync_metadata.fields.sync_metadata$fn__47016$update_metadata_BANG___47021$fn__47022.invoke(sync_metadata.clj:95)"
"sync.sync_metadata.fields.sync_metadata$fn__47016$update_metadata_BANG___47021.invoke(sync_metadata.clj:89)"
"sync.sync_metadata.fields$fn__47081$sync_and_update_BANG___47086$fn__47087.invoke(fields.clj:80)"
"sync.sync_metadata.fields$fn__47081$sync_and_update_BANG___47086.invoke(fields.clj:72)"
"sync.sync_metadata.fields$fn__47109$sync_fields_for_table_BANG___47118$fn__47124$fn__47126.invoke(fields.clj:102)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:149)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"sync.sync_metadata.fields$fn__47109$sync_fields_for_table_BANG___47118$fn__47124.invoke(fields.clj:90)"
"sync.sync_metadata.fields$fn__47109$sync_fields_for_table_BANG___47118.invoke(fields.clj:83)"
"sync.sync_metadata.fields$fn__47154$sync_fields_BANG___47159$fn__47160$iter__47161__47165$fn__47166$fn__47167.invoke(fields.clj:117)"
"sync.sync_metadata.fields$fn__47154$sync_fields_BANG___47159$fn__47160$iter__47161__47165$fn__47166.invoke(fields.clj:116)"
"sync.sync_metadata.fields$fn__47154$sync_fields_BANG___47159$fn__47160.invoke(fields.clj:116)"
"sync.sync_metadata.fields$fn__47154$sync_fields_BANG___47159.invoke(fields.clj:111)"
"sync.util$fn__38107$run_step_with_metadata__38112$fn__38116$fn__38118.invoke(util.clj:356)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invokeStatic(util.clj:104)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invoke(util.clj:98)"
"sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:121)"
"sync.util$with_start_and_finish_debug_logging.invoke(util.clj:118)"
"sync.util$fn__38107$run_step_with_metadata__38112$fn__38116.invoke(util.clj:353)"
"sync.util$fn__38107$run_step_with_metadata__38112.invoke(util.clj:348)"
"sync.util$fn__38299$run_sync_operation__38304$fn__38305$fn__38306.invoke(util.clj:441)"
"sync.util$fn__38299$run_sync_operation__38304$fn__38305.invoke(util.clj:441)"
"sync.util$fn__38299$run_sync_operation__38304.invoke(util.clj:435)"
"sync.sync_metadata$fn__47841$sync_db_metadata_BANG___47846$fn__47847$fn__47848.invoke(sync_metadata.clj:51)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:149)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:147)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"driver.mongo$eval2022$fn__2023$f__825__auto____2024.invoke(mongo.clj:73)"
"driver.mongo.util$_with_mongo_connection$fn__818.invoke(util.clj:118)"
"util.ssh$with_ssh_tunnel_STAR_.invokeStatic(ssh.clj:106)"
"util.ssh$with_ssh_tunnel_STAR_.invoke(ssh.clj:91)"
"driver.mongo.util$_with_mongo_connection.invokeStatic(util.clj:97)"
"driver.mongo.util$_with_mongo_connection.invoke(util.clj:92)"
"driver.mongo$eval2022$fn__2023.invoke(mongo.clj:72)"
"sync.util$sync_in_context$fn__38013.invoke(util.clj:140)"
"sync.util$with_db_logging_disabled$fn__38010.invoke(util.clj:131)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invokeStatic(util.clj:104)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invoke(util.clj:98)"
"sync.util$with_start_and_finish_logging$fn__37999.invoke(util.clj:116)"
"sync.util$with_sync_events$fn__37994.invoke(util.clj:90)"
"sync.util$with_duplicate_ops_prevented$fn__37985.invoke(util.clj:69)"
"sync.util$do_sync_operation.invokeStatic(util.clj:168)"
"sync.util$do_sync_operation.invoke(util.clj:165)"
"sync.sync_metadata$fn__47841$sync_db_metadata_BANG___47846$fn__47847.invoke(sync_metadata.clj:50)"
"sync.sync_metadata$fn__47841$sync_db_metadata_BANG___47846.invoke(sync_metadata.clj:47)"
"task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:39)")

May 04 23:51:24 ERROR metabase.sync.util :: Error running sync step: Output of sync-fields-for-table! does not match schema: 

 (not (map? a-java.sql.BatchUpdateException)) 


("--> sync.sync_metadata.fields$fn__47109$sync_fields_for_table_BANG___47118.invoke(fields.clj:83)"
"sync.sync_metadata.fields$fn__47154$sync_fields_BANG___47159$fn__47160$iter__47161__47165$fn__47166$fn__47167.invoke(fields.clj:117)"
"sync.sync_metadata.fields$fn__47154$sync_fields_BANG___47159$fn__47160$iter__47161__47165$fn__47166.invoke(fields.clj:116)"
"sync.sync_metadata.fields$fn__47154$sync_fields_BANG___47159$fn__47160.invoke(fields.clj:116)"
"sync.sync_metadata.fields$fn__47154$sync_fields_BANG___47159.invoke(fields.clj:111)"
"sync.util$fn__38107$run_step_with_metadata__38112$fn__38116$fn__38118.invoke(util.clj:356)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invokeStatic(util.clj:104)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invoke(util.clj:98)"
"sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:121)"
"sync.util$with_start_and_finish_debug_logging.invoke(util.clj:118)"
"sync.util$fn__38107$run_step_with_metadata__38112$fn__38116.invoke(util.clj:353)"
"sync.util$fn__38107$run_step_with_metadata__38112.invoke(util.clj:348)"
"sync.util$fn__38299$run_sync_operation__38304$fn__38305$fn__38306.invoke(util.clj:441)"
"sync.util$fn__38299$run_sync_operation__38304$fn__38305.invoke(util.clj:441)"
"sync.util$fn__38299$run_sync_operation__38304.invoke(util.clj:435)"
"sync.sync_metadata$fn__47841$sync_db_metadata_BANG___47846$fn__47847$fn__47848.invoke(sync_metadata.clj:51)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:149)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:147)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"driver.mongo$eval2022$fn__2023$f__825__auto____2024.invoke(mongo.clj:73)"
"driver.mongo.util$_with_mongo_connection$fn__818.invoke(util.clj:118)"
"util.ssh$with_ssh_tunnel_STAR_.invokeStatic(ssh.clj:106)"
"util.ssh$with_ssh_tunnel_STAR_.invoke(ssh.clj:91)"
"driver.mongo.util$_with_mongo_connection.invokeStatic(util.clj:97)"
"driver.mongo.util$_with_mongo_connection.invoke(util.clj:92)"
"driver.mongo$eval2022$fn__2023.invoke(mongo.clj:72)"
"sync.util$sync_in_context$fn__38013.invoke(util.clj:140)"
"sync.util$with_db_logging_disabled$fn__38010.invoke(util.clj:131)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invokeStatic(util.clj:104)"
"sync.util$with_start_and_finish_logging_SINGLEQUOTE_.invoke(util.clj:98)"
"sync.util$with_start_and_finish_logging$fn__37999.invoke(util.clj:116)"
"sync.util$with_sync_events$fn__37994.invoke(util.clj:90)"
"sync.util$with_duplicate_ops_prevented$fn__37985.invoke(util.clj:69)"
"sync.util$do_sync_operation.invokeStatic(util.clj:168)"
"sync.util$do_sync_operation.invoke(util.clj:165)"
"sync.sync_metadata$fn__47841$sync_db_metadata_BANG___47846$fn__47847.invoke(sync_metadata.clj:50)"
"sync.sync_metadata$fn__47841$sync_db_metadata_BANG___47846.invoke(sync_metadata.clj:47)"
"task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:39)")
```

Unfortunately I don't know how to reproduce this but would be more than happy to field questions or try and help ID the cause.  I can tell you that the field it is referring too in the query (`on`) isn't a real field in the DB.  Not sure where it is getting that from.   We do have fields labeled `processedOn`.

**Comment by camsaul** [comment]
Fixed by #9920

**Comment by stieg** [comment]
Sweet.  Thanks @camsaul
